packages/core/src/tracing/instrumentation/cloud/aws/s3.js

```diff
@@ -0,0 +1,215 @@
+};
+
+exports.activate = function activate() {
+  isActive = true;
+};
+
+exports.deactivate = function deactivate() {
+  isActive = false;
+};
+
+function instrumentAWS(AWS) {
+  shimmer.wrap(AWS.Service.prototype, 'makeRequest', shimMakeRequest);
+  shimmer.wrap(AWS.Request.prototype, 'promise', shimRequestPromise);
+}
+
+function shimMakeRequest(originalMakeRequest) {
+  return function() {
+    if (isActive) {
+      const originalArgs = new Array(arguments.length);
+      for (let i = 0; i < originalArgs.length; i++) {
+        originalArgs[i] = arguments[i];
+      }
+
+      return instrumentedMakeRequest(this, originalMakeRequest, originalArgs);
+    }
+
+    return originalMakeRequest.apply(this, arguments);
+  };
+}
+
+function instrumentedMakeRequest(ctx, originalMakeRequest, originalArgs) {
+  /**
+   * Original args for ALL AWS SDK Requets: [method, params, callback]
+   */
+
+  if (!isOperationShimmable(originalArgs[0]) || typeof originalArgs[2] !== 'function') {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  if (cls.tracingSuppressed()) {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  const parentSpan = cls.getCurrentSpan();
+
+  if (!parentSpan || isExitSpan(parentSpan)) {
+    return originalMakeRequest.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(() => {
+    const span = cls.startSpan('s3', EXIT);
+    span.ts = Date.now();
+    span.stack = tracingUtil.getStackTrace(instrumentedMakeRequest);
+    span.data.s3 = buildS3Data(originalArgs[0], originalArgs[1]);
+
+    const _originalCallback = originalArgs[2];
+
+    originalArgs[2] = cls.ns.bind(function(err, data) {
+      finishSpan(err, data, span);
+      setImmediate(() => finishSpan(err, data, span));
+      return _originalCallback.apply(this, arguments);
+    });
+
+    return originalMakeRequest.apply(ctx, originalArgs);
+  });
+}
+
+function shimRequestPromise(originalPromise) {
+  return function shimmedPromise() {
+    if (isActive && isOperationShimmable(this.operation)) {
+      const originalArgs = new Array(arguments.length);
+      for (let i = 0; i < originalArgs.length; i++) {
+        originalArgs[i] = arguments[i];
+      }
+
+      return instrumentedRequestPromise(this, originalPromise, originalArgs);
+    }
+    return originalPromise.apply(this, arguments);
+  };
+}
+
+function instrumentedRequestPromise(ctx, originalPromise, originalArgs) {
+  if (cls.tracingSuppressed()) {
+    return originalPromise.apply(ctx, originalArgs);
+  }
+
+  if (!isOperationShimmable(ctx.operation)) {
+    return originalPromise.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(function() {
+    const span = cls.startSpan('s3', EXIT);
+    span.ts = Date.now();
+    span.stack = tracingUtil.getStackTrace(instrumentedRequestPromise);
+    const _promise = originalPromise;
+
+    originalPromise = cls.ns.bind(function patchedPromise() {
+      span.data.s3 = buildS3Data(this.operation, this.params);
```

Comment:
I would have expected `ctx.operation` here, not `this.operation`. Same with `this.params`. To what does `this` refer here?